cartographer/mapping_2d/local_trajectory_builder.h

```diff
@@ -43,6 +43,7 @@ class LocalTrajectoryBuilder {
   struct InsertionResult {
     std::shared_ptr<const mapping::TrajectoryNode::Data> constant_data;
     std::vector<std::shared_ptr<const Submap>> insertion_submaps;
+    sensor::RangeData inserted_range_data;
```

Comment:
This should not be added here. We want all the data from `last_pose_estimate_` to be made available via the callback, even if there is not `InsertionResult`. This is redundant then. Moreover, we should then remove `pose_estimate()` from the global trajectory builder interface.

Reply:
Ooops. I completely forgot that I can't pass the range data if the insertion result was a null pointer.

Reply:
Ooops. I forgot what I was doing here last week.

Reply:
About the contents of pose_estimate: I want the original range data just like was previously stored inside a trajectory node, not a transformed point cloud. Can I change contents of  pose_estimate struct to reflect this?

Reply:
One additional thing: sometimes `nullptr` is returned because there isn't enough accumulated data. How will the global trajectory builder know not to invoke the callback, because `last_pose_estimate_` hasn't been updated in the meantime and it would not make sense to invoke the callback with the same data? Maybe by comparing `time` from the `PoseEstimate`?

Reply:
True. Maybe we should then get rid of the `last_post_estimate_` completely and return this information instead, possible containing an insertion result, i.e. either nullptr, the `MatchingResult` not containing an `InsertionResult`, or the `MatchingResult` including an `InsertionResult`? Then it would also be clearer that these are results. `pose_estimate()` seems weird since it would only be called after adding some range data anyway.

Reply:
Ok. I will get on it. To clarify: `AddRangeData` should return a `std::unique_ptr<MatchingResult>` (which is `nullptr` if accumulation is incomplete); if `AddAccumulatedRangeData` is called, the returned `MatchingResult` contains the accumulated range data, the pose estimate (plus the gravity alignment quaternion for 2D) and time, and a `std::unique_ptr<InsertionResult>`: `nullptr` if motion filter has eliminated the accumulated range data, or the `InsertionResult` as before. `NodeId` remains the second argument of the callback.

Reply:
Why the gravity alignment? I thought of the contents of `PoseEstimate` which contain the full 3D pose in 2D and 3D. Other than that: Yes.

Reply:
If we pass `gravity_aligned_range_data` to the callback, along with `pose_estimate`, we might need  `gravity_alignment.inverse()` to correctly interpret the `gravity_aligned_range_data` with respect to some internal parts of Cartographer. E.g. `gravity_aligned_range_data` is inserted into a submap like this: 
```
submap.InsertRangeData(
      TransformRangeData(gravity_aligned_range_data,
                         transform::Embed3D(pose_estimate_2d.cast<float>()))
```
where the relationship between `pose_estimate_2d` and `pose_estimate` is the following:
`pose_estimate = transform::Embed3D(pose_estimate_2d) * gravity_alignment;`

So, you would need to multiply `pose_estimate` with `gravity_alignment.inverse()` in order to be able to exactly reconstruct how the range data was (or would be) inserted into a submap.

This is not a problem with scans which are inserted into the SPG, because `gravity_alignment` is stored into the corresponding node's `constant_data,` and you can retrieve it from there. But you would not be able to do this for scans which had been discarded by the motion filter, because there would be no corresponding trajectory node.

I am not saying that I need this, I am just being cautious. We can drop it, if you prefer.

Reply:
@wohe before we go on addressing other things you mentioned, pleave have a look at what I wrote above.

Reply:
1. We never stored unfiltered data anywhere. Can you explain what you mean by that?
2. Passing the range data in different frames depending on whether it is 2D or 3D is a bad idea as it means that code using the data has to implement separate code path to handle both. There is also no benefit when changing this. The user can gets both the range data and the pose and can decide to transform it as the user wishes.

Reply:
1. That's fine, keeping the range data not just the point cloud sounds like a good idea. But why not transform? I was thinking of the voxel filtering which happens before, i.e. the range data is a subset of the input point cloud.
2. Why not do as before:
```
TransformRangeData(gravity_aligned_range_data,
                         transform::Embed3D(pose_estimate_2d.cast<float>()))
```
? Then the range data is in the local map frame. If you are later interested in the range data in tracking frame, you can multiply by `pose_estimate.inverse()`.

Reply:
Okay, we can do that. I did not like that at first because it was not following the way how the filtered range data was stored in the node constant data (not transformed to the local frame).

Reply:
As I said the important part is to be consistent across 2D and 3D. The other option is obviously to transform the 2D range data to be in tracking frame.

Reply:
This would correspond to `gravity_aligned_range_data * pose_estimate_2d gravity_alignment.inverse() * pose_estimate_2d.inverse()`, right?

```
range_data_in_local = gravity_aligned_range_data * pose_estimate_2d

pose_estimate = pose_estimate_2d * gravity_alignment
pose_estimate.inverse() = 
    gravity_alignment_inverse() * pose_estimate_2d.inverse()

range_data_in_tracking = range_data_in_local * pose_estimate.inverse() = 
    gravity_aligned_range_data * pose_estimate_2d gravity_alignment.inverse() 
    * pose_estimate_2d.inverse()
```

And it would make sense to apply the same to filtered `RangeData` stored in node `constant_data`, right?

Reply:
From your explanation it sounds you should keep it like it is. If there is not a good reason why something should be changed, we should not change it. This feels like bikeshedding otherwise.

Regarding the transform: Points get multiplied to the very right, not the left. Your last equality is wrong. `pose_estimated_2d` is from gravity-aligned to local map frame. `gravity_alignment` from tracking to gravity-aligned frame, so:

1. `gravity_aligned_range_data` transformed by `pose_estimate_2d` is in local map frame.
2. `gravity_aligned_range_data` transformed by `gravity_alignment.inverse()` is in tracking frame.

Also: Please do not change `constant_data`. It contains the data in the suitable format for loop closure detection and is furthermore completely unrelated to this PR.

Reply:
Okay. I am going to introduce `range_data_in_local`, pass it to `InsertIntoSubmap` by const reference, and then `move` it into `MatchingResult`.